Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp

```diff
@@ -0,0 +1,163 @@
+ */
+void CreateMonteCarloWorkspace::init() {
+  auto mustBePositive = std::make_shared<Mantid::Kernel::BoundedValidator<int>>();
+  mustBePositive->setLower(0);
+
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("InputWorkspace", "", Direction::Input),
+                  "Input Workspace containing data to be simulated");
+  declareProperty("Seed", 32, mustBePositive,
+                  "Integer that seeds initialises the random-number generator, for reproducibility");
+  declareProperty("MonteCarloEvents", 0, mustBePositive,
+                  "Number of Monte Carlo events to simulate. Defaults to integral of input workspace if 0.");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("OutputWorkspace", "", Direction::Output),
+                  "Name of output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         int seedInput,
+                                                                                         API::Progress &progress) {
+
+  Mantid::HistogramData::HistogramY outputY(cdf.size(), 0.0);
+  std::mt19937 gen(seedInput);
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+
+  int progressInterval = std::max(1, numIterations / 100); // Update progress every 1%
+
+  for (int i = 0; i < numIterations; ++i) {
+    double randomNum = dis(gen);
+    auto it = std::lower_bound(cdf.begin(), cdf.end(), randomNum);
+    size_t index = std::distance(cdf.begin(), it);
+
+    if (index < outputY.size()) {
+      outputY[index] += 1.0;
+    }
+
+    if (i % progressInterval == 0) {
+      progress.report("Generating random data...");
+    }
+  }
+  return outputY;
+}
+
+/**
+ *  Compute a normalized CDF [0..1] from the given histogram data.
+ */
+std::vector<double> CreateMonteCarloWorkspace::computeNormalizedCDF(const Mantid::HistogramData::HistogramY &yData) {
+  std::vector<double> cdf(yData.size());
+  std::partial_sum(yData.begin(), yData.end(), cdf.begin());
+  double totalCounts = cdf.back();
+
+  if (totalCounts > 0.0) {
+    // Normalize the CDF
+    std::transform(cdf.begin(), cdf.end(), cdf.begin(), [totalCounts](double val) { return val / totalCounts; });
+  } else {
+    g_log.warning("Total counts are zero; normalization skipped.");
+  }
+  return cdf;
+}
+
+/**
+ *  Determine how many iterations to use for MC sampling.
+ *  If userMCEvents > 0, use that directly; otherwise use the integral of the input data.
+ */
+int CreateMonteCarloWorkspace::integrateYData(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
```

Comment:
Think you missed a variable here when converting to camel case sorry for not picking up on that in my previous reviews (maybe I do too much in python!)
```suggestion
  double totalCounts = std::accumulate(yData.begin(), yData.end(), 0.0);
```

Reply:
Thanks for letting me know. I fixed it. I also squashed a lot of the commits to make it easier to follow